contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -308,17 +308,75 @@ response reduces the complexity of these clients.
+represented using a list/slice of objects, where each condition has a similar
+structure. This collection should be treated as a map with a key of `type`.
+
+Conditions are most useful when they follow some consistent conventions:
+
+* Conditions should be added to explicitly convey properties that users and
+  components care about rather than requiring those properties to be inferred
+  from other observations.  Once defined, the meaning of a Condition can not be
+  changed arbitrarily - it becomes part of the API, and has the same backwards-
+  and forwards-compatibility concerns of any other part of the API.
+
+* Controllers should apply their conditions to a resource the first time they
+  visit the resource, even if the `status` is Unknown. This allows other
+  components in the system to know that the condition exists and the controller
+  is making progress on reconciling that resource.
+
+   * Not all controllers will observe the previous advice about reporting
+     "Unknown" or "False" values. For known conditions, the absence of a
+     condition status should be interpreted the same as `Unknown`, and
+     typically indicates that reconciliation has not yet finished (or that the
+     resource state may not yet be observable).
+
+* For some conditions, `True` represents normal operation, and for some
+  conditions, `False` represents normal operation. ("Normal-true" conditions
+  are sometimes said to have "positive polarity", and "normal-false" conditions
+  are said to have "negative polarity".) Without further knowledge of the
+  conditions, it is not possible to compute a generic summary of the conditions
+  on a resource.
+
+* Condition type names should make sense for humans; neither positive nor
+  negative polarity can be recommended as a general rule. A negative condition
+  like "MemoryExhausted" may be easier for humans to understand than
+  "SufficientMemory". Conversely, "Ready" or "Succeeded" may be easier to
+  understand than "Failed", because "Failed=Unknown" or "Failed=False" may
+  cause double-negative confusion.
+
+* Condition type names should describe the current observed state of the
+  resource, rather than describing the current state transitions. This
+  typically means that the name should be an adjective ("Ready", "OutOfDisk")
+  or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb
+  ("Deploying"). Intermediate states may be indicated by setting the status of
+  the condition to `Unknown`.
+
+  * For state transitions which take a long period of time, use a condition
+    which indicates that the transition succeeded ("Provisioned" or
+    "SizeAllocated") with an intermediate state of `Unknown`. Note that this
+    pattern makes an explicit distinction between "condition not set" (which
+    means the reconciler hasn't seen the latest state) and "condition set to
+    Unknown" (which means that the reconciler has seen the state but hasn't
+    finished the reconciliation).
```

Comment:
I guess your goal is that some watcher can observe progress and know that it's done? You should document how long after the operation the condition should be left there.

But I'm not convinced this is the way to do it, what if there are two resizes back to back? I don't think you can really have conditions 1:1 with operations.

I think "Resizing=true" that is present during the resize and removed after the resize is also a legitimate--and possibly less confusing--way to represent this. Seeing "ResizingSucceeded=Unknown" is disconcerting, even if it's technically accurate. 

Maybe the way to strengthen your point is to talk about how the failure of the operation will be represented. I think you can't really do that with a "Reszing" condition, but if you have another way of representing failure, it's OK. If you need a way to represent failure, then "ResizingSucceeded=false" works better.

Reply:
Actually, my concern is that the watcher can determine whether progress is being made in a generic way. (Particularly in the presence of CRDs and a variety of types, detecting "controller is stuck" vs "backend system is slow" can be quite difficult.)

I'd suggest not removing the condition when the operation is done (hence, `Provisioned` or `SizeAllocated`). The goal is twofold:

1. Presence of the condition along with `LastTransitionTime` helps to debug whether the controller has seen the most recent `spec`.

1. Ability to signal success and failure of these operations in a generic way that will stick with the resource seems useful for the "hey, could you help debug this with me" case -- a resize failed condition that disappears after 2 hours might be gone by the time an expert takes a look at the correct resource in the cluster.


As for the human ergonomics, I was hoping that the `Reason` and `Message` fields would suffice:

```yaml
status:
  conditions:
  - name: Resizing
    status: True
    Reason: FileSystemExtensionInProgress
    Message: The system is currently extending the filesystem on the disk after resize.
```

```yaml
status:
  conditions:
  - name: SizeAvailable
    status: Unknown
    Reason: FileSystemExtensionInProgress
    Message: The system is currently extending the filesystem on the disk after resize.
```

(Note that I changed this to `SizeAvailable` because there's probably a distinction between allocation failures and filesystem extension failures.)

Reply:
Maybe a way to say this is something like "conditions should not be transient, they should more or less always be on the object with the appropriate value".

But this doesn't rule out something like "Resizing", you can just set it False when the object is the correct size. At first it seems like people would need to look at the Reason to know if it's a "good" false (no resize needed) or a "bad" false (resize failed), but in the latter case we can just leave it in "resizing" (Reason is e.g. "stuck", message is "resize operation can't complete due to error foo").

Reply:
That works for me -- my concern was more specifically that the set of signalled Conditions not be transient, which this addresses. PTAL!